REF: simplify info.py

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Enable polymorphism and builder pattern in ``info.py``.
The main benefits:
- Separated data representation from data itself
- More clear construction of each element via builder pattern
- Different builders for various cases (verbose with counts, verbose without counts, non-verbose) enabled one eliminate numerous if statements

May be useful to implement the present PR first and then build ``Series.info`` (https://github.com/pandas-dev/pandas/pull/31796) on top of this.
I do realize that I deleted ``BaseInfo`` class.
Now I see that it is required by the referenced PR, so I can move it back.

**Note:** needed to change test behavior. Here is the reason why.
I noticed inconsistency:
```
>>> df = pd.DataFrame({'long long column': np.random.rand(1000000)})
>>> df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 1 columns):
 #   Column            Non-Null Count    Dtype
---  ------            --------------    -----
 0   long long column  1000000 non-null  float64
dtypes: float64(1)
memory usage: 7.6 MB
```
Here we have two spaces between columns.

However, if we create a dataframe with 10000 columns, then the distance between # col and "Column" is only one space.
```
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame(np.random.rand(3, 10001))
>>> with open('out.txt', 'w') as buf: df.info(verbose=True, buf=buf)
```

```
$ tail out.txt
 9993  9993    float64
 9994  9994    float64
 9995  9995    float64
 9996  9996    float64
 9997  9997    float64
 9998  9998    float64
 9999  9999    float64
 10000 10000   float64
dtypes: float64(10001)
memory usage: 234.5 KB
```
See, only one space between the first and the second columns.
I find it inconsistent.
So, I changed it, so that there are two spaces everywhere.
What do you think?

I am going to finalize it by documenting and introducing typing where required.

ivanovmg committed:
REF: polymorphism and builder pattern in info

ivanovmg committed:
TST: adjust expected gap between # and Column

ivanovmg committed:
LINT: remove TYPE_CHECKING import

For some reason I got import error when importing Series
under TYPE_CHECKING only.
So, now I import Series normally and delete TYPE_CHECKING

ivanovmg committed:
TST: add test for empty dataframe info

ivanovmg committed:
FIX: handle empty dataframe case

ivanovmg committed:
REF: extract abstract cls BaseInfo and builder

ivanovmg committed:
DOC: fix docstrings

ivanovmg commented:
I separated ``BaseInfo`` and ``TableBuilderAbstract`` to make integration of series info smoother.
It seems that all that is required for the series info is to derive ``SeriesInfo`` from ``BaseInfo`` and ``SeriesTableBuilder`` from ``TableBuilderAbstract``. Moreover, some of the methods from ``DataFrameTableBuilder`` can probably be useful for ``SeriesTableBuilder`` as well (like column widths detection, for instance). Or al least they can be similar.

MarcoGorelli commented on pandas/io/formats/info.py:
```diff
@@ -209,147 +229,315 @@ def info(self) -> None:
+    HEADERS = [
+        " # ",
+        "Column",
+        "Dtype",
```
can you remove the trailing comma so it fits on one line?

ivanovmg replied:
Did that. But IMHO, it is more clear, when it is multilne. Anyway, did as you suggested.

MarcoGorelli commented on pandas/io/formats/info.py:
```diff
@@ -72,92 +70,114 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
+
     def __init__(
         self,
         data: FrameOrSeries,
```
:thinking: not sure if `FrameOrSeries` is correct here, as the comment near it says
```
# FrameOrSeriesUnion  means either a DataFrame or a Series. E.g.
# `def func(a: FrameOrSeriesUnion) -> FrameOrSeriesUnion: ...` means that if a Series
# is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
# in, either a DataFrame or a Series is returned.
```
, which `__init__` doesn't return anything. However, `FrameOrSeriesUnion` also talks about either a Series or a DataFrame being returned, so I'm not sure how to type this if nothing is returned... @topper-123 any suggestions?

ivanovmg replied:
Looking at the documentation, I guess ``FrameOrSeries`` means the following.
Consider ``x: FrameOrSeries``.
Function ``f(x) -> FrameOrSeries``.
Let us say, that now ``x`` is a dataframe, then ``f(x)`` must return ``DataFrame``, but not ``Series``.
Meanwhile ``FrameOrSeriesUnion`` allows that ``f(x)`` returns different type.

MarcoGorelli replied:
Yes, that's my understanding too, although technically `__init__` doesn't return anything, so...just `Union[Series, DataFrame]`?

ivanovmg replied:
``FrameOrSeriesUnion = Union["DataFrame", "Series"]``, I can use it.

It we use ``FrameOrSeries``, then it does not mean that we must return.
It just means that **if** we return, we **must** return the same type.

It is not like I object against changes.
I only mean that these are not required.
mypy would raise issues if any.

ivanovmg replied:
I tried ``FrameOrSeriesUnion`` and it leads to mypy complains.

```
pandas/io/formats/info.py:140: error: Incompatible return value type (got "Union[Any, ExtensionDtype]", expected "Series")  [return-value]
pandas/io/formats/info.py:201: error: Incompatible return value type (got "Union[Any, ExtensionDtype]", expected "Series")  [return-value]
```

So, I leave it as it is for now.

MarcoGorelli replied:
I think you're right, e.g. `pandas/io/pytables.py` contains:
```python
def to_hdf(
    path_or_buf,
    key: str,
    value: FrameOrSeries,
    mode: str = "a",
    complevel: Optional[int] = None,
    complib: Optional[str] = None,
    append: bool = False,
    format: Optional[str] = None,
    index: bool = True,
    min_itemsize: Optional[Union[int, Dict[str, int]]] = None,
    nan_rep=None,
    dropna: Optional[bool] = None,
    data_columns: Optional[Union[bool, List[str]]] = None,
    errors: str = "strict",
    encoding: str = "UTF-8",
):
```

Let's keep as is then, thanks for explaining!

MarcoGorelli commented on pandas/io/formats/info.py:
```diff
@@ -72,92 +70,114 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
+        self.memory_usage = self._initialize_memory_usage(memory_usage)
+
+    def _initialize_memory_usage(
+        self,
```
I don't think `self` is used here, can this be a static method or a module-level function?

ivanovmg replied:
I made it a static method.

MarcoGorelli commented on pandas/io/formats/info.py:
```diff
@@ -209,147 +229,315 @@ def info(self) -> None:
+        " # ",
+        "Column",
+        "Non-Null Count",
+        "Dtype",
```
likewise can this trailing commas be removed?

MarcoGorelli commented on pandas/tests/io/formats/test_info.py:
```diff
@@ -87,7 +101,7 @@ def test_info_verbose():
 
     res = buf.getvalue()
-    header = " #    Column  Dtype  \n---   ------  -----  "
+    header = " #     Column  Dtype  \n---    ------  -----  "
```
why has this changed?

ivanovmg replied:
I changed this test because of the inconsistency with the column spacing I observed.
See "Note: needed to change test behavior. Here is the reason why." in the PR description.

ivanovmg replied:
The spacing between the columns will not change in general. This particular test has 10000 columns, which makes the first column wide. As a result, there is an inconsistent spacing between the columns. I forced it to be 2 spaces regardless of the column width as I suppose it is reasonable.

MarcoGorelli replied:
> See "Note: needed to change test behavior. Here is the reason why." in the PR description.

Thanks, I should've noticed that. I think your explanation and reasoning makes sense, though perhaps it would be better to have a separate PR to fix this, and then keep this one to enable polymorphism?

ivanovmg replied:
OK, I will create the issue and make a small PR to fix it. But the present PR will basically overwrite this fix anyway.

ivanovmg replied:
Created PR #36766.

MarcoGorelli reviewed (changes requested):
Nice, this looks much better than what I'd tried!

ivanovmg committed:
REF: de-duplicate line_numbers, columns, dtypes

ivanovmg committed:
CLN: make HEADERS one-liners

ivanovmg committed:
REF: make _initialize_memory_usage staticmethod

ivanovmg committed:
DOC/TYP: add docstrings and type annotations

ivanovmg committed:
REF: create memory usage string in BaseInfo

WillAyd commented on pandas/io/formats/info.py:
```diff
@@ -72,92 +82,123 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
+        return self.dtypes.value_counts().groupby(lambda x: x.name).sum()
 
-    def info(self) -> None:
+    def to_buffer(self, *, buf, max_cols, verbose, null_counts) -> None:
```
Can you annotate these arguments?

WillAyd commented on pandas/io/formats/info.py:
```diff
@@ -209,147 +250,355 @@ def info(self) -> None:
         %(examples_sub)s
         """
-        lines = []
+        printer = InfoPrinter(
```
Hmm at first glance I think this delegation pattern is a little confusing but maybe I am overlooking something. What is the advantage to calling `DataFrameInfo.to_buffer` from `.info` instead of calling `InfoPrinter.to_buffer` instead?

ivanovmg replied:
I call ``DataFrameInfo.to_buffer`` from ``DataFrame`` because of the following reasons.
1. Mainly did not want to mess with the docstring (which is defined in ``DataFrameInfo.to_buffer``)
2. I wanted to avoid import of ``InfoPrinter`` into ``pandas/core/frame.py``
3. Probably it is a good idea to make public API function as disconnected from the internal API as possible. Right now changes inside ``pandas/io/formats/io.py`` will have slight effect on ``DataFrame`` class, since the info printing functionality is encapsulated inside ``DataFrameInfo``. If, however, we call ``InfoPrinter.to_buffer`` from ``DataFrame.info()``, then we would need to ensure that all the possible changes made inside ``InfoPrinter`` and ``DataFrameInfo`` are properly communicated to ``DataFrame.info``. I hope it makes sense.

But I do not strongly object against having ``InfoPrinter.to_buffer`` called from ``DataFrame.info()``.

ivanovmg committed:
TYP: annotate DataFrameInfo.to_buffer

ivanovmg committed:
REF: rename kwarg null_counts -> show_counts

Changes concern internal implementation only.
Public API remains the same.

ivanovmg committed:
CLN: reuse exceeds_info_cols in show_counts

ivanovmg committed:
REF: extract property exceeds_info_rows

ivanovmg committed:
DOC: add docstrings in InfoPrinter

ivanovmg committed:
CLN: rename counts -> dtype_counts

This is more reasonable name for
mapping {dtype: number of columns with this dtype}.

ivanovmg committed:
REF: extract property non_null_counts

ivanovmg committed:
REF: extract method _get_non_null_counts

ivanovmg committed:
REF: generate rows for performance improvement

Previously there was a need to iterate over all columns
multiple times to create strcols.
This commit creates row data by iterating
over the dataframe columns only once.
Then zip(*self.strrows) is used to form strcols.

ivanovmg committed:
REF: remove COL_SPACE class attribute

ivanovmg committed:
CLN: drop attribute strcols

ivanovmg committed:
CLN: simplify iteration, without unpacking

ivanovmg committed:
CLN: rename _get_* gen functions to _gen_*

ivanovmg committed:
CLN: avoid unpacking in for loop

ivanovmg committed:
REF: replace setter with initializer function

ivanovmg committed:
REF: eliminate dep of TableBuilder on InfoPrinter

ivanovmg committed:
CLN: use yield from in _gen_rows

ivanovmg committed:
DOC: add docstrings to abstract properties

ivanovmg committed:
Merge branch 'master' into refactor/info-polymorphism

jreback commented on pandas/io/formats/info.py:
```diff
@@ -72,92 +83,147 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
+        self.data = data
+        self.memory_usage = self._initialize_memory_usage(memory_usage)
+
+    @staticmethod
```
don't use staticmethods can be  a class method or an instance method or a module level function

ivanovmg replied:
Moved to the module level. My concern here is that this method is necessary only for the instantiation/validation of ``self.memory_usage``, so would not it be more reasonable to keep it close to the actual usage position... Anyway.

jreback commented on pandas/io/formats/info.py:
```diff
@@ -72,92 +83,147 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
-            for system-level memory consumption, and include it in the returned
-            values.
+    @property
+    def mem_usage(self) -> int:
```
would rather not abbeviate this method (so maybe have to use _memory_usage for the storage

ivanovmg replied:
I renamed it to ``memory_usage_bytes``.

jreback commented on pandas/io/formats/info.py:
```diff
@@ -209,147 +275,353 @@ def info(self) -> None:
-        counts = dtypes.value_counts().groupby(lambda x: x.name).sum()
-        collected_dtypes = [f"{k[0]}({k[1]:d})" for k in sorted(counts.items())]
-        lines.append(f"dtypes: {', '.join(collected_dtypes)}")
+    def _select_verbose_table_builder(self) -> Type["DataFrameTableBuilderVerbose"]:
```
can you fold this into _select_table_builder, there is too much indirection here.

ivanovmg replied:
Did that.

jreback commented on pandas/io/formats/info.py:
```diff
@@ -209,147 +275,353 @@ def info(self) -> None:
+        """Product in a form of list of lines (strings)."""
+
+
+class DataFrameTableBuilder(TableBuilderAbstract):
```
can you reduce the builder abstraction a bit. this is too java-esque. you can just pass parameterize / use partial functions. I find this style too different from the rest of pandas. let's make this simpler.

ivanovmg replied:
As you requested, I dropped ``DataFrameTableBuilderVerboseNoCounts`` and ``DataFrameTableBuilderVerboseWithCounts``.
Instead, I parametrized ``DataFrameTableBuilderVerbose`` (with_counts = True/False). I personally prefer the original implementation, but this one is also OK. My concerns with the new implementation are rather minor. Two if statements checking for ``with_counts`` in two separate functions (fragile), and presence of ``_gen_non_null_counts``, which is required for the table with counts and is irrelevant for table without counts.

I leave ``TableBuilderAbstract`` as @MarcoGorelli is working on series info, thus it will be reasonable for the builders to have the same interface.

MarcoGorelli replied:
Regarding Series.info - if you want to work on it I'm happy for you to take it over (else I'll go back to it once this PR is in)

ivanovmg replied:
@MarcoGorelli, I can work on Series.info from the code point of view. Since you already have documentation and tests, it seems to be quite manageable for me.

MarcoGorelli replied:
Cool, I'll close that PR then! If for whatever reason you change your mind, please ping me and I'll take it up again

ivanovmg replied:
Good. I will start working on series info once this is PR merged.

ivanovmg committed:
REF: remove method for verbose builder selection

The logic with the selection of verbose builder was moved
to function _select_table_builder.
This makes the logic quite more complicated though.

ivanovmg committed:
REF: parametrize verbose table builders

Previously there were two separate classes for verbose table builder.
 - DataFrameTableBuilderVerboseWithCounts
 - DataFrameTableBuilderVerboseNoCounts

The review considered it as a complex pattern.
The present commit leaves DataFrameTableBuilderVerbose,
makes is parametrized (with_counts = True/False).

ivanovmg committed:
REF: move static method to module level function

ivanovmg committed:
CLN: rename mem_usage -> memory_usage_bytes

ivanovmg committed:
DOC: add docstrings to iterator funcs

ivanovmg committed:
Merge branch 'master' into refactor/info-polymorphism

ivanovmg committed:
PERF: property gross_column_widths to attribute

ivanovmg committed:
TYP: attribute gross_column_widths

ivanovmg committed:
Merge branch 'master' into refactor/info-polymorphism

ivanovmg committed:
Merge branch 'master' into refactor/info-polymorphism

ivanovmg commented:
Merged master.
By the way, looks like the recent bug (https://github.com/pandas-dev/pandas/issues/37245) would have emerged in this implementation.

ivanovmg commented:
CI checks failures related to datetime. Looks like these are common across multiple PRs now.

ivanovmg commented:
@jreback ping

jreback commented:
> Merged master.
> By the way, looks like the recent bug (#37245) would not have emerged in this implementation.

how is that?

jreback commented:
@MarcoGorelli if you have any further comments.

I am ok with this as is. @ivanovmg really prefer to have much smaller / incremental PRs. Sure i understand once in a while its advantageous to do it all at once. But review will inevitable take *way* longer.

ivanovmg commented:
> 
> 
> > Merged master.
> > By the way, looks like the recent bug (#37245) would not have emerged in this implementation.
> 
> how is that?

Probably because in this PR we iterate over items, rather than access by index (in master).

MarcoGorelli commented:
> @MarcoGorelli if you have any further comments.

I can review during the weekend, but there's no need to wait for me if you'd like this in sooner

ivanovmg commented:
> 
> 
> > @MarcoGorelli if you have any further comments.
> 
> I can review during the weekend, but there's no need to wait for me if you'd like this in sooner

I could be ready with series.info by the end of this week.
This PR is holding me.
I guess, we can figure out the remaining concerns (if any) in that one (with series info).

jreback commented:
thanks @ivanovmg
